### DISCUSSION TITLE
test weirdness

### DISCUSSION THREAD

**Discussion by tessus** [orhun/rustypaste#263] [General]
I was trying to look into the test `test_upload_duplicate_file` that fails with tarpaulin. Turns out this test also fails with [`nextest`](https://nexte.st/).

It is the only test that creates an upload dir `fs::create_dir(test_upload_dir)?;`

When I tried to use the current dir (as all other tests do), this test stalls and runs for about 90 seconds - and passes (in nextest, I haven't tried tarpaulin).

It hangs for 40s at both instances of `test::call_service(....)`.

With the upload dir set to something, it doesn't stall but the test fails. I am very puzzled by this. @orhun are you aware of any side effects of `actix_web::test::test_utils`?

**Comment by orhun** [comment]
That's really weird. I'm not aware of any issues with `test_utils` but maybe it is an concurrency issue? Something might be locking?

**Reply by tessus** [reply]
But I don't understand how it could be a concurrency issue. The test is basically isolated. The name of the uploaded file is different than in any other test. And then you just send that request twice. Not even in parallel, but in sequence.

**Reply by orhun** [reply]
Yeah, I don't really know what's up with it. I can take a look again soon.

**Reply by orhun** [reply]
I'm trying locally and the test passes for me:

```
$ cargo nextest run -- "test_upload_duplicate_file"
    Finished test [unoptimized + debuginfo] target(s) in 0.11s
    Starting 1 test across 2 binaries (37 skipped)
        PASS [   0.011s] rustypaste server::tests::test_upload_duplicate_file
------------
     Summary [   0.012s] 1 test run: 1 passed, 37 skipped
```

**Reply by tessus** [reply]
Hmm, interesting:

```
$ cargo nextest run
    Finished test [unoptimized + debuginfo] target(s) in 0.18s
    Starting 38 tests across 2 binaries
        PASS [   0.390s] rustypaste config::tests::test_get_tokens
        PASS [   0.388s] rustypaste config::tests::test_parse_deprecated_config
        PASS [   0.383s] rustypaste mime::tests::test_match_mime_type
        PASS [   0.384s] rustypaste header::tests::test_expiry_date
        PASS [   0.385s] rustypaste header::tests::test_content_disposition
        PASS [   0.386s] rustypaste file::tests::test_file_checksum
        PASS [   0.387s] rustypaste config::tests::test_space_handling
        PASS [   0.389s] rustypaste config::tests::test_parse_config
        PASS [   0.391s] rustypaste auth::tests::test_extract_tokens
        PASS [   0.821s] rustypaste paste::tests::test_paste_data
        PASS [   0.697s] rustypaste random::tests::test_generate_url
        PASS [   0.439s] rustypaste server::tests::test_delete_file
        PASS [   0.476s] rustypaste server::tests::test_auth
        PASS [   0.385s] rustypaste server::tests::test_delete_file_without_token_in_config
        PASS [   0.317s] rustypaste server::tests::test_index
        PASS [   0.257s] rustypaste server::tests::test_index_with_landing_page
        PASS [   0.196s] rustypaste server::tests::test_index_with_landing_page_file
        PASS [   0.148s] rustypaste server::tests::test_index_with_landing_page_file_not_found
        PASS [   0.500s] rustypaste server::tests::test_list
        PASS [   0.426s] rustypaste server::tests::test_payload_limit
        PASS [   0.348s] rustypaste server::tests::test_upload_expiring_file
        FAIL [   0.388s] rustypaste server::tests::test_upload_duplicate_file

--- STDOUT:              rustypaste server::tests::test_upload_duplicate_file ---

running 1 test
test server::tests::test_upload_duplicate_file ... FAILED

failures:

failures:
    server::tests::test_upload_duplicate_file

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 37 filtered out; finished in 0.00s


--- STDERR:              rustypaste server::tests::test_upload_duplicate_file ---
Error: Os { code: 17, kind: AlreadyExists, message: "File exists" }

   Canceling due to test failure: 8 tests still running
        PASS [   0.309s] rustypaste server::tests::test_upload_file
        PASS [   0.272s] rustypaste server::tests::test_upload_file_override_filename
        PASS [   0.236s] rustypaste server::tests::test_upload_oneshot
        PASS [   0.191s] rustypaste server::tests::test_upload_oneshot_url
        PASS [   0.576s] rustypaste server::tests::test_upload_remote_file
        PASS [   0.527s] rustypaste server::tests::test_upload_same_filename
        PASS [   0.435s] rustypaste server::tests::test_version
        PASS [   0.898s] rustypaste server::tests::test_list_expired
        PASS [   0.478s] rustypaste server::tests::test_upload_url
        PASS [   0.255s] rustypaste util::tests::test_get_expired_files
        PASS [   0.382s] rustypaste server::tests::test_version_without_auth
        PASS [   0.318s] rustypaste server::tests::test_version_without_config
        PASS [   0.153s] rustypaste util::tests::test_safe_join_path
        PASS [   0.234s] rustypaste util::tests::test_glob_match
        PASS [   0.146s] rustypaste util::tests::test_sha256sum
        PASS [   0.141s] rustypaste util::tests::test_system_time
------------
     Summary [   1.748s] 38 tests run: 37 passed, 1 failed, 0 skipped
        FAIL [   0.388s] rustypaste server::tests::test_upload_duplicate_file
error: test run failed
```

**Reply by tessus** [reply]
Which still makes no sense, because the test uses its own upload dir.

**Reply by orhun** [reply]
I think you should use `--test-threads 1` to avoid possible clashes. That's what I do in the CI as well.

Can you also try `cargo nextest run -- "test_upload_duplicate_file"`?

**Reply by tessus** [reply]
Yes, I always use `--test-threads 1` with `cargo test`. However, `nextest` does tests in parallel and none are failing except `test_upload_duplicate_file`.

Running `cargo nextest run -- "test_upload_duplicate_file"` works and passes for me.

It still does not explain the situation. All other tests use the current dir, which means there should be an issue, or with port conflicts, but apparently the actix-tests do not use tcpip or they made it work, so that we can run tests in parallel even though all of them use the same port - see another thing that makes no sense. Otherwise every test with `nextest` would have to fail.

**Reply by orhun** [reply]
yeah, it's pretty weird. Maybe it is worth reporting to `nextest` then? - maybe that would lead to some useful discussion about why this is happening

**Reply by tessus** [reply]
I already did, but nothing useful so far.

https://github.com/nextest-rs/nextest/issues/1369

**Reply by tessus** [reply]
also, it seems we have issues again with the #246 PR.

**Reply by orhun** [reply]
oops, I missed that, just watched the issue.

> also, it seems we have issues again with the https://github.com/orhun/rustypaste/pull/246 PR.

god damn it!

I will take a look soon :D

**Reply by tessus** [reply]
this time another test fails. What worries me is that codecov and nextest might point out that there is actually an issue with the test, but we haven't noticed it.
But I checked the tests and they seem perfectly fine from a test perspective: process, reason, test case, result. Unless I am missing something.

**Reply by tessus** [reply]
@orhun, I've got an answer from the tarpulin dev that could help:

https://github.com/xd009642/tarpaulin/issues/1488#issuecomment-2080450739

**Reply by orhun** [reply]
@tessus thanks for tracking this!

Should we give it a shot for reverting to `cargo-tarpaulin` in CI and using the suggested flags then?
